ENT-9608: Wait for service lifecycle event before start token loading and handle result set re-sync

adelel1 opened the pull request:
Changes made:

Now wait for either ServiceLifecycleEvent.BEFORE_STATE_MACHINE_START or ServiceLifecycleEvent.STATE_MACHINE_STARTED before we start tokens loading into memory. By waiting for the above events token loading takes place after the vault is initialised.

When reading vault states into query pages if we detect that a page has become out of sync, then we reread the last page and determine if we can sync with that page, and so on until we find a page we can sync with. We establish sync points by slightly overlapping the database query.

Regarding the 2 ignored test methods. The VaultWatcherService now calls the register method of AppServiceHub. The test uses a AppServiceHub mock defined in Corda, unfortunately in the mock method this throws a not implemented exception. Which is the exception we get when test method is run. Couldn't define my own mock as the mock implementation adds the VaultWatcherService to Corda service list. One solution would be to update Corda to not throw exceptions when not implemented (by the mock). 

Had to revert having having the vault changes run on the UPDATER thread. Tests started to fail. Problem is a new token is created, this then causes the event to be queued to run in the UPDATER thread. But the next line of code in the test selects from the VaultWatcherService so is assuming the VaultWatcherService has been updated (which it may not have been, as the update is occurring in a different thread now), so causing failure.

adelel1 committed:
ENT-9608: Wait for service lifecycle event before start token loading. This means it will start after vault is initialised.
Also handles result set becoming out of sync, due to paralell removal of tokens.

adelel1 committed:
ENT-9608: Ignore 2 tests, as needs update to Corda in createMockCordaService

adelel1 committed:
ENT-9608: Fix failing test, insufficient balance.

adelel1 committed:
ENT-9608: Revert using UPDATER thread to handle vault changes.

adelel1 committed:
ENT-9608L Reducing token numbers in tests.

rick-r3 commented on modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt:
```diff
@@ -111,36 +124,131 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
+                    LOG.info("Token loading has become out of sync, will re-sync")
+                    return -1
+                }
+                private fun getNewPageSpecification(): PageSpecification {
```
Is a better name to imply what it does something like `popPriorPageSpecification` ?

adelel1 replied:
Yes, changed.

rick-r3 commented on modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt:
```diff
@@ -111,36 +124,131 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
+                    return specification.pageSize * (specification.pageNumber-1) + listEntryIndex
+                }
+
+                private fun toListEntryIndex(resultSetIndex: Int, specification: PageSpecification): Int {
```
`toIndexWithinPage` ?

adelel1 replied:
Changed.

rick-r3 commented on modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt:
```diff
@@ -111,36 +124,131 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
+
+                    syncPoints.forEach {
+                        val listIndex = toListEntryIndex(it.lastKnownResultSetIndex, specification)
+                        if (listIndex in lowResultSetIndex..highResultSetIndex && states[listIndex] == it.stateAndRef) {
```
This comparison of `listIndex` (index within page) to result set indexes doesn't make any sense.  Surely it doesn't work.   I'd be tempted to remove this `if` and just always fall back to the `states.contains(it.stateAndRef)` (or however it ends up written) as less conditions to write tests for.

adelel1 replied:
Oops. Now fixed by just checking the index is in range. I would like to keep this if in. As hopefully it would be the normal case. i.e. when nothing has changed in the list we don't do any extra searching etc. So performance wise should be the same as now.

rick-r3 commented on modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt:
```diff
@@ -186,7 +294,7 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
                 val existingMark = __backingMap.putIfAbsent(stateAndRef, PLACE_HOLDER)
                 existingMark?.let {
-                    LOG.warn("Attempted to overwrite existing token ${stateAndRef.ref}, this suggests incorrect vault behaviours")
+                    LOG.warn("Attempted to overwrite existing token ${stateAndRef.ref}, this suggests a result set re-sync occurred")
```
We should reconsider this warning.  Does something break when this condition occurs?  Should the function return or re-do the indexing (as it does now)?   Should it really be a `debug`?

adelel1 replied:
I've changed to a debug. It is just saying that a duplicate entry has been inserted. This would be because of a resync. Although in the code i do try to minimise the duplicate inserts by moving to the next unread entry (in all cases except one, where we have seen the whole list).

rick-r3 commented on modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt:
```diff
@@ -111,36 +124,131 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
             }
             return TokenObserver(emptyList(), uncheckedCast(vaultObservable), ownerProvider, asyncLoader)
         }
+        data class SyncPoint(val stateAndRef: StateAndRef<FungibleToken>, val specification: PageSpecification, var lastKnownResultSetIndex: Int)
```
Maybe `lastKnownResultSetIndex` should be `resultSetIndexOfStateAndRef` ?

rick-r3 replied:
Should also be a `Long`

adelel1 replied:
Changed to a long. Have changed to lastKnownResultSetIndexOfStateAndRef
I do want to keep the lastKnown in the name because if the result set has changed, then this value would be inaccurate, but would be the 'last known' value.

rick-r3 commented on modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt:
```diff
@@ -111,36 +124,131 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
+                    }
+                }
+
+                private fun toResultSetIndex(listEntryIndex: Int, specification: PageSpecification): Int {
```
I think we should be using `Long` for the return value here.  i.e. in theory it is `Int * Int` so could exceed the size of `Int`.   Then need to ripple through other data structures.

adelel1 replied:
Changed.

adelel1 committed:
ENT-9608: Update following review comments:
  Now do a binary search for the highest state ref less than one we have seen.
  Made sync points a single value now.

rick-r3 commented on modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt:
```diff
@@ -111,36 +125,165 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
+                    return (resultSetIndex - ((specification.pageNumber-1) * specification.pageSize)).toInt()
+                }
+
+                fun seachForLargestSeenStateRef(values: List<StateAndRef<FungibleToken>>, element: StateAndRef<FungibleToken>): Int {
```
Use `Collections.binarySearch()` ?  Otherwise we need to unit test this to death.

adelel1 replied:
Done.

rick-r3 commented on modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt:
```diff
@@ -111,36 +125,165 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
+                    return result
+                }
+                private fun compare(first: StateAndRef<FungibleToken>, second: StateAndRef<FungibleToken>): Int {
+                    if (first.ref.txhash.bytes.toHexString() < second.ref.txhash.bytes.toHexString()) {
```
`SecureHash` is `Comparable`.  So you can use Kotlin `compareValues(first.ref.txhash, second.ref.txhash)`

adelel1 replied:
Done. Using compareBy now.

rick-r3 commented on modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt:
```diff
@@ -111,36 +125,165 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
+                    if (first.ref.txhash.bytes.toHexString() > second.ref.txhash.bytes.toHexString()) {
+                        return 1
+                    }
+                    if (first.ref.index < second.ref.index) {
```
And then if the secure hashes are equal (compareValues == 0), use `compareValues(first.ref.index, second.ref.index)`

adelel1 replied:
Done. Using compareBy now.

rick-r3 reviewed (commented):
Looks thematically correct.  Think you can use more built in stuff for the binary search and comparison.

adelel1 committed:
ENT-9608: Switched to using Collections.binarySearch.
